Automattic/woocommerce-subscriptions-core
# Refactor our Related Orders Data Store classes to support WC CRUD methods

Fixes #145

## Description

This PR updates our `WCS_Related_Order_Store_Cached_CPT` and `WCS_Related_Order_Store_CPT` classes to use WC CRUD methods to be able to support both WP Posts and High-Performance Order Storage (HPOS) environments.

> **Note** - Doesn't include renaming our classes
>
> This PR doesn't rename our classes to remove the reference to Custom Post Types with the `_CPT` suffix. Renaming these classes and deprecating the class isn't a critical part to this refactor, so I will create a separate issue to tackle this at a later time.

> **Note** - Can't test with HPOS
>
> This PR does not update our `WCS_Post_Meta_Cache_Manager` class which triggers a lot of the functionality in these related orders store classes, this means you cannot test this PR with HPOS enabled. You must be using WP Posts.

### What are these caches?

Subscription related orders (renewals, switch and resubscribe orders) record their relationship in order meta. Historically finding these subscription related orders was costly as it required querying the database for orders with specific meta (a meta query and post join). To fix this in WC Subscriptions 2.3 we introduced a persistent caching layer. In subscription meta we now store a single key to keep track of the subscriptions related orders. 

<img width="548" alt="Screen Shot 2022-11-09 at 4 46 27 pm" src="https://user-images.githubusercontent.com/8490476/200758656-4b3f762e-a571-4203-aadc-a2c28d0e8aeb.png">

Now to get a subscription's list of renewal orders we simply just refer to this single meta key on the subscription rather than multiple values stored on every order. 

These cache classes act as gatekeepers by providing APIs for fetching related orders `get_related_order_ids()` and they also house all the hooks for keeping the cache up to date. If a renewal order is trashed, this class is responsible for removing that record for the subscription's related order cache. 

There are 2 public facing pages that include some details of these classes: 

- [**Subscription Caches** → **_Subscription Related Order Cache_**](https://woocommerce.com/document/subscriptions/develop/cache/#section-1)
- [**What’s New in Subscriptions 2.3** → **_Performance_**](https://woocommerce.com/document/version-2-3/#section-7)

### What's new?

There's a fair bit changing in this PR so let's go through it:
 - New Subscriptions data store functions:
   - `delete_all_metadata_by_key` - Previously we used WP's `delete_metadata()` to remove every cache meta we had stored. There's no equivalent in the WC HPOS data stores so we created our own.
   - ~`get_metadata_by_key` - When fetching related order cache meta data we need to make sure it's fetched directly from the database. This function pulls the metadata directly from the new orders table or if using wp posts, it calls `get_post_meta()`~
 - Replaced instanced of  `get_post_meta`, `update_post_meta`  & `delete_post_meta` with relevant WC CRUD method.
 - Replaced uses of `get_posts()` with new `wcs_get_orders_with_meta_query()`
 - ~New helper function `get_subscription_with_cache_meta()` to help us get an instance of WC Subscriptions with all of our related order cache meta in memory. This solves a problem I was facing with trying to fetch the cache returning an empty array and trying to update the cache, creating a duplicate postmeta entry.~

## How to test this PR

Testing this PR should hopefully be pretty straightforward and the main thing we need to test is making sure the subscription's related-order cache is properly updated when its list of related orders changes, and when an order is updated we also need to make sure the relevant related order meta is updated on the orders.

Flows to test:
1. After processing a renewal order, the renewal order is visible in the related orders table on the Edit Subscription page
2. After processing a renewal order, the new renewal order ID is inside the `_subscription_renewal_order_ids_cache` on the subscription
3. After processing a renewal order, make sure the renewal order has the `_subscription_renewal` and proper subscription ID saved in postmeta.
4. After trashing a renewal order, the renewal order ID is removed from the `_subscription_renewal_order_ids_cache` array in subscription postmeta
5. After trashing a renewal order, the renewal order should not show up in the related orders table on the subscription

## Product impact
